myia/unity.py

```diff
@@ -0,0 +1,503 @@
+        return f"UnionVar({self.tag}, {self.values})"
+
+
+class RestrictedVar(Var[U]):
+    """Variable restricted to a set of values."""
+
+    __slots__ = ('legal_values',)
+
+    def __init__(self, tag: str,
+                 legal_values: Tuple[Value[U], ...]) -> None:
+        """Create a RestrictedVar."""
+        super().__init__(tag)
+        self.legal_values = legal_values
+
+    @classmethod
+    def _parse_args(cls, args: Tuple[Any, ...], kwargs: Dict[Any, Any]):
+        assert len(kwargs) == 0
+        assert len(args) == 2
+        assert isinstance(args[0], str)
+        return (args[0], tuple(sorted(args[1], key=id)))
+
+    def matches(self, value: Value[U]) -> bool:
+        """Return True if the variable matches the value."""
+        if isinstance(value, RestrictedVar):
+            return all(v in self.legal_values for v in value.legal_values)
+        return value in self.legal_values
+
+    def __repr__(self) -> str:
+        return f"RestrictedVar({self.tag}, {self.legal_values})"
+
+
+class FilterVar(Var[U]):
+    """Variable restricted to values that pass a filter function."""
+
+    __slots__ = ('filter',)
+
+    def __init__(self, tag: str, filter: FnFiltT[U]) -> None:
+        """Create a FilterVar."""
+        super().__init__(tag)
+        self.filter = filter
+
+    @classmethod
+    def _parse_args(cls, args: Tuple[Any, ...], kwargs: Dict[Any, Any]):
+        assert len(kwargs) == 0
+        assert len(args) == 2
+        assert isinstance(args[0], str)
+        return args
+
+    def matches(self, value: Value[U]) -> bool:
+        """Return True if the variable matches the value."""
+        if isinstance(value, RestrictedVar):
+            return all(self.filter(v) for v in value.legal_values)
+        if isinstance(value, FilterVar):
+            return self.filter == value.filter
+        return self.filter(value)
+
+    def __repr__(self) -> str:
+        return f"FilterVar({self.tag}, {self.filter})"
+
+
+def expandlist(lst: Iterable[Value[U]]) -> List[Value[U]]:
+    """Flatten the Seq instances in a sequence."""
+    lst = list(lst)
+    off = 0
+    for i, e in enumerate(list(lst)):
+        if isinstance(e, Seq):
+            lst[off + i:off + i + 1] = e
+            off += len(e) - 1
+    return lst
+
+
+def noseq(fn: Callable[[Value[U]], Value[U]], u: Value[U]) -> Value[U]:
+    """Make sure that there are no Seq in the value."""
+    um = fn(u)
+    if isinstance(um, Seq):
+        raise TypeError("Multiple values in single-value position")
+    return um
+
+
+class Unification(Generic[T]):
+    """Unification engine."""
+
+    def __init__(self) -> None:
+        """Create a unification engine."""
+        self._seq = 1
+        self._vars: Dict[Any, Var[T]] = dict()
+
+    def _get_var(self, cls: Type[Var[T]], args: Tuple[Any, ...]) -> Var[T]:
+        pargs = cls._parse_args(args, dict())
+        if pargs not in self._vars:
+            self._vars[pargs] = cls(*pargs)
```

Comment:
I wasn't sure what the utility of caching `Var` instances is, and after talking with Arnaud we determined that it wasn't truly useful and a leftover of earlier design decisions. So I think we should just drop the cache and create `Var` instances directly.

Reply:
The cache is gone.